spec/index.md

```diff
@@ -5033,6 +5101,215 @@ S with
+    details = LoadSnapshot {
+      module_hash = SHA-256(New_state.wasm_module);
+    };
+  };
+}
+
+liquid_balance(
+  New_balance,
+  New_reserved_balance,
+  freezing_limit(
+    S.compute_allocation[A.canister_id],
+    S.memory_allocation[A.canister_id],
+    S.freezing_threshold[A.canister_id],
+    memory_usage_wasm_state(New_state.wasm_state) +
+      memory_usage_raw_module(New_state.raw_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(S.snapshots[A.canister_id]),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+if S.memory_allocation[A.canister_id] > 0:
+  memory_usage_wasm_state(New_state.wasm_state) +
+    memory_usage_raw_module(New_state.raw_module) +
+    memory_usage_canister_history(New_canister_history) ≤ S.memory_allocation[A.canister_id]
+
+```
+
+State after  
+
+```html
+
+S with
+    canisters[A.canister_id] = {
+      wasm_state = S.snapshots[A.canister_id].wasm_state;
+      raw_module = S.snapshots[A.canister_id].raw_module;
+      module = parse_wasm_mod(S.snapshots[A.canister_id].raw_module;
+      chunk_store = S.snapshots[A.canister_id].chunk_store;
+      public_custom_sections = parse_public_custom_sections(S.snapshots[A.canister_id].raw_module);
+      private_custom_sections = parse_private_custom_sections(S.snapshots[A.canister_id].raw_module);
+    }
+    certified_data[A.canister_id] = S.snapshots[A.canister_id].certified_data
+    balances[A.canister_id] = New_balance
+    reserved_balances[Canister_id] = New_reserved_balance
+    canister_history[Canister_id] = New_canister_history
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin;
+        response = Reply (candid());
+        refunded_cycles = M.transferred_cycles;
+      }
+
+```
+
+#### IC Management Canister: List canister snapshots
+
+Only the controllers of the given canister can get a list of the existing snapshots.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'list_canister_snapshots'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+```
+
+State after
+
+```html
+
+S with
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin
+        response = Reply (candid(S.snapshots[A.canister_id]))
+        refunded_cycles = M.transferred_cycles
+      }
+
+```
+#### IC Management Canister: Delete canister snapshot
+
+A snapshot may be deleted only by the controllers of the canister for which the snapshot was taken.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'delete_canister_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+S.snapshots[A.canister_id] = (deleted)
```

Comment:
this belongs to State after; instead we should check here that the snapshot stored for the canister has a matching snapshot ID according to A.snapshot_id